### PR TITLE
KEYCLOAK-14817 Allow JS adapter to be bundled as ES module

### PR DESCRIPTION
**Pull request by jonkoops**
This pull request introduces a compilation step for the JavaScript adapter with a couple of design goals:

## Introducing ECMAScript modules

Modern JavaScript tooling such as Webpack increasingly prefer that packages are provided as [ECMAScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) in order to properly optimize the final distribution. For example, compiling `keycloak-js` into an Angular application will produce the following warning:

A compilation warning for keycloak js provided by Angular CLI

To mitigate this problem the source code has been updated to make use of the ECMAScript module syntax. These files are then compiled into both a version that uses ES modules, and a version that uses the traditional CommonJS format for NodeJS environments and older bundlers.

This is made possible by using the `main` field in the package to refer to the CommonJS module which is compiled to the `lib/` directory, and using the `module` field to refer to the modern module which is compiled to the `es/` directory. More information about this pattern can be found in the [Webpack documentation](https://webpack.js.org/guides/author-libraries/#final-steps).

Another advantage of this approach is that is no longer necessary to inline minified versions of dependencies, instead we can rely on the official versions from NPM. 

## Splitting `keycloak-js` and `keycloak-authz` into seperate packages

Currently `keycloak-authz` is included in the `keycloak-js` package, but is not part of the root of the package resulting in the following very akward import statement:

```ts
import KeycloakAuthorization from 'keycloak-js/dist/keycloak-authz'
```
This is not only ugly, but also undocumented behaviour. In the new situation this code will receive it's own separate NPM package called `keycloak-authz`. This simplifies the import as well and makes things a lot cleaner:

```ts
import KeycloakAuthorization from 'keycloak-authz'
```  

This change is not purely cosmetic either, since we are shipping two versions of the package bundled together to support the different module types it is needed to provide a single entry point to the package. To facilitate splitting these packages and possibly also introduce more packages (such as the adapters for Cordova and Cordova Native) [workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces) are used to ensure they can be integrated together.

**Note:** It might be a good idea to put these packages together into an [organisation](https://docs.npmjs.com/organizations) on NPM. Example the packages could published as follows:

- `keycloak-js` ➡️ `@keycloak/client`
- `keycloak-authz`  ➡️  `@keycloak/authorization`

## Compiling the static scripts from the source code

The JavaScript adapter can be used in two ways, either by including it as a script served statically from the Keycloak server,  or by installing the package from NPM. This has worked so far, but using a single source file for both is not without complications, such as always including all dependencies in-line by copy pasting the source. 

In this PR the package is compiled from source by means of the TypeScript compiler, this is the version will be published to NPM. A separate Webpack build will then compile this package to a set of scripts which are backwards compatible and can be served from the Keycloak server (these same files will end up in the JAR).

The advantage of this approach is that we can work on and improve the code without having to worry about the traditional constraints of having everything in a single statically served file without sacrificing backwards compatibility.

## Improving maintainability

Since this PR introduces the TypeScript compiler it is now possible to start splitting up the over 1500 lines of code that make up the adapter into smaller chunks that we can test individually. And since we are using the TypeScript compiler we can also start taking advantage of TypeScript in order to improve correctness of the code itself. This will greatly improve maintainability and allows us to start moving the unit tests in as well, with the end of goal of eventually putting the source into it's own standalone repository.

## Improved typings

The types for `keycloak-js` and `keycloak-authz` have been improved. Both the default exports `Keycloak` and `KeycloakAuthorization` now use the `class` type instead of the `function` type, allowing developers to use the `new` keyword to create instances. The global namespaces have also been dropped, instead relying on explicit imports to further emphasize the difference between using the package over the 'global' static version.

## Familiar tooling for contributors

One of the things that scares away potential contributors to the JavaScript client is that in order to build it there is a heavy reliance on Maven. Although this is great for Java developers it is far from ideal for developers less experienced with this tooling.

This change introduces familiar tooling to front-end developers such as the TypeScript compiler and Webpack, which have become an industry standard.

## Other notable changes
- Frontend plugin has been updated to the latest version
- Latest Node (version 16) is now used throughout the project
- Release version tagging now happens directly in the release script
- `base64-js` has been updated to the latest version
- Dependabot config has been added to keep dependencies up to date

### CODE REVIEW

**Comment by jonkoops** [comment]
@stianst would it be possible to get some eyes on this from the core team?

**Comment by vmuzikar** [comment]
@jonkoops Sorry, I haven't reviewed but. But this seems to me like a bigger architectural change. Was this discussed somewhere if it's even something we want to do? I'm especially interested in @stianst @ssilvert and @edewit opinions.

**Comment by jonkoops** [comment]
@vmuzikar This is certainly a large change indeed. I've discussed it before with @stianst before on the mailing list and various tickets, but not quite in this detail of course.

The intention here is to allow the source code for the adapter to be slowly refactored to TypeScript and create test coverage in the adapter itself so it can eventually possibly be lifted into a standalone repository. This is impossible at the moment as the testing suite is tightly coupled to the rest of the product.

Consider this PR a working draft proposal on what we'd like to do next with the JavaScript client. I'd like to collect feedback from the core team on this approach.

**Comment by edewit** [comment]
I love this PR and I'm :+1: on moving this out of the keycloak repo so we can attract more javascript developers to have eyes on this.

File: adapters/oidc/js/package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  "private": true,
+  "workspaces": [
+    "packages/keycloak-js",
+    "packages/keycloak-authz"
+  ],
+  "engines": {
+    "node": "^14"
+  },
+  "scripts": {
+    "bundle": "rimraf target/classes && cross-env TS_NODE_PROJECT=\"tsconfig.webpack.json\" webpack"
```
Comment (edewit) [inline adapters/oidc/js/package.json:11]: Wouldn't it be nice to add a build script here as well? So you can build all the packages from one place?
```suggestion
    "build": "yarn workspaces run build",
    "bundle": "rimraf target/classes && cross-env TS_NODE_PROJECT=\"tsconfig.webpack.json\" webpack"
```

**Review by edewit — DISMISSED** [review]

File: adapters/oidc/js/package.json (hunk 1 above)
Comment (jonkoops) [reply adapters/oidc/js/package.json:12]: Yeah, I can see how that could be useful. I've added the script and updated the pom file to reflect this as well.

**Comment by DaSchTour** [comment]
> I love this PR and I'm 👍 on moving this out of the keycloak repo so we can attract more javascript developers to have eyes on this.

I would then love to help with refactoring to TypeScript.

File: adapters/oidc/js/.browserslistrc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+IE 11
```
Comment (jonkoops) [inline adapters/oidc/js/.browserslistrc:1]: This seemed like a good target for browser support for the statically server adapter. I'd like to have some kind of official acknowledgement of this though.

**Comment by jonkoops** [comment]
I've added some changes to add the correct version tags to the the packages before publishing. @stianst this should take care of adding the proper versions at release time. However we might want to add the ability to add the correct pre-release version to [keycloak-bot](https://github.com/keycloak/keycloak/commits?author=keycloak-bot) as well.

**Comment by jonkoops** [comment]
The unit test are failing. I honestly do not have enough knowledge on the tooling on the Java side understand what is going wrong here let alone fix it. Any assistance here would be much appreciated.

**Comment by jonkoops** [comment]
I've fixed all of the tests and the CI is looking green (one exception, but seems to be the same on master). I believe this PR is now in good enough shape to be reviewed and possibly merged.

**Comment by jonkoops** [comment]
I've pushed some changes to keep this PR fresh and resolve some merge conflicts. I've also made the following changes:
- Node 16 is now used everywhere instead of Node 14
- Since Node 16 has NPM version 7 which supports workspaces Yarn has been removed
- A Dependabot config has been added to keep dependencies up to date

**Review by edewit — APPROVED** [review]

**Comment by ssilvert** [comment]
We've put this off so I'm closing it for now.  @jonkoops will definitely reopen it later though.

**Pull request closed without merging**